File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> arr = new ArrayList<>();

        for(int i =0;i<nums.length;i++){
            if(i > 0 && nums[i] == nums[i-1]){
                continue;
            }

            int l = i + 1;
            int r = nums.length - 1;

            while(l < r){
                int sum = nums[i] + nums[l] + nums[r];

                if(sum == 0){
                   List<Integer> arr2 = new ArrayList<>();
                    arr2.add(nums[i]);
                    arr2.add(nums[l]);
                    arr2.add(nums[r]);
                    arr.add(arr2);
                    l++;
                    r--;
                    while(l < r && nums[l] == nums[l - 1]) l++;
                    while(l < r && nums[r] == nums[r + 1]) r--;
                }else if(sum > 0){
                    r--;
                }else{
                    l++;
                }
            }
        }

        return arr;
    }
}

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        nums.sort()
        lists = []

        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i-1]:
                continue   

            l = i + 1
            r = len(nums) - 1

            while l < r:
                sum = nums[i] + nums[l]+ nums[r]
                if sum == 0:
                    nelist = [nums[i],nums[l],nums[r]]
                    lists.append(nelist)
                    l+=1
                    r-=1
                    while l < r and nums[l] == nums[l-1]: l+=1
                    while r >= l and nums[l] == nums[r+1]: r-=1
                elif sum < 0:
                    l+=1    
                else:
                    r-=1

        return lists                